src/audio/pipeline/pipeline-graph.c

```diff
@@ -232,13 +240,20 @@ void pipeline_disconnect(struct comp_dev *comp, struct comp_buffer *buffer, int
 	 * written back, overwriting the modified header. FIXME: this is still a
 	 * problem with different cores.
 	 */
-	if (comp_list != buf_list->next)
-		dcache_writeback_invalidate_region(uncache_to_cache(buf_list->next),
-						   sizeof(struct list_item));
-	if (comp_list != buf_list->prev)
-		dcache_writeback_invalidate_region(uncache_to_cache(buf_list->prev),
-						   sizeof(struct list_item));
+	buffers_after_exist = comp_list != buf_list->next;
+	buffers_before_exist = comp_list != buf_list->prev;
+	needs_sync_prev = uncache_to_cache(buf_list->prev);
+	needs_sync_next = uncache_to_cache(buf_list->next);
+	if (buffers_after_exist)
+		dcache_writeback_region(needs_sync_next, sizeof(struct list_item));
+	if (buffers_before_exist)
+		dcache_writeback_region(needs_sync_prev, sizeof(struct list_item));
+	/* buffers before or after can be prefetched here */
 	list_item_del(buf_list);
+	if (buffers_after_exist)
+		dcache_invalidate_region(needs_sync_next, sizeof(struct list_item));
+	if (buffers_before_exist)
+		dcache_invalidate_region(needs_sync_prev, sizeof(struct list_item));
```

Comment:
I now see your point @lyakh that maybe it would be easier to have everything cached :D 

This does kind of scream that this would need to be abstracted somewhere. Basicly it's problematic to have the list members in the buffer struct, and basicly they cannot be used without doing the full dance above.

It still seems this indeed needs to be done, so I'm ok to merge this and then try to improve the abstraction/interface in follow-ups.

Reply:
@kv2019i yes, I have a follow-up PR prepared after this one to put this in a separate function. You still have a function with some pretty complex cache sync, but at least it's separate... And I then reuse those two functions for buffer.c and module_adapter.c